pubtools/_quay/security_manifest_pusher.py

```diff
@@ -0,0 +1,535 @@
+                List of push items.
+            target_settings (dict):
+                Target settings.
+        """
+        self.push_items = push_items
+        self.target_settings = target_settings
+        # TODO: change this to whatever is decided
+        self.cosign_private_key_path = self.target_settings["cosign_private_key_path"]
+        self.cosign_public_key_path = self.target_settings["cosign_public_key_path"]
+        self.quay_host = self.target_settings.get("quay_host", "quay.io").rstrip("/")
+
+        self._src_quay_client = None
+        self._dest_quay_api_client = None
+
+    @property
+    def src_quay_client(self):
+        """Create and access QuayClient for source image."""
+        if self._src_quay_client is None:
+            self._src_quay_client = QuayClient(
+                self.target_settings["source_quay_user"],
+                self.target_settings["source_quay_password"],
+                self.target_settings.get("source_quay_host") or self.quay_host,
+            )
+        return self._src_quay_client
+
+    @property
+    def dest_quay_api_client(self):
+        """Create and access QuayApiClient for dest image."""
+        if self._dest_quay_api_client is None:
+            self._dest_quay_api_client = QuayApiClient(
+                self.target_settings["dest_quay_api_token"], self.quay_host
+            )
+        return self._dest_quay_api_client
+
+    def cosign_get_security_manifest(self, image_ref: str, output_file: str) -> bool:
+        """
+        Use cosign to get security manifest from an image and save it to a file.
+
+        Args:
+            image_ref (str):
+                Image reference to get the manifest from.
+            output_file (str):
+                File where to save the manifest.
+        Returns (bool):
+            Whether the manifest was gathered successfully (True) or not (False).
+        """
+        cmd = ["cosign", "download", "sbom", image_ref, "--output-file", output_file]
+        LOG.info(f"Running command '{' '.join(cmd)}'")
+        result = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True)
+
+        if result.returncode:
+            LOG.warning(f"Command {' '.join(cmd)} has failed: {result.stdout}")
+            return False
+
+        return True
+
+    def cosign_get_existing_attestation(self, image_ref: str, output_file: str) -> bool:
+        """
+        Use cosign to verify and get an attestation, if it exists.
+
+        Args:
+            image_ref (str):
+                Image reference to get the manifest from.
+            output_file (str):
+                File where to save the attestation.
+        Returns (bool):
+            Whether the attestation was gathered successfully (True) or not (False).
+        """
+        cmd = [
+            "cosign",
+            "verify-attestation",
+            "--key",
+            self.cosign_public_key_path,
+            image_ref,
+            "--output-file",
```

Comment:
I wonder if there's a reason why stdout cannot be used instead of file

Reply:
I think it can, but should it?

Reply:
I personally prefer solutions which doesn't produce mess. For example: cosign_get_existing_attestation takes output file as attribute it's called in merge_and_push_security_manifest but I don't see any cleanup there.

Reply:
The entire workflow happens inside a TemporaryDirectory (see method push_item_security_manifests). The directory along with its contents is deleted once the code is outside of its context.